dynamiqs/qarrays/utils.py

```diff
@@ -80,3 +179,220 @@ def stack(qarrays: Sequence[QArray], axis: int = 0) -> QArray:
+    """  # noqa: E501
+    _warn_qarray_dims(x, dims)
+
+    if isinstance(x, Qobj):
+        return x
+    elif isinstance(x, DenseQArray):
+        return _dense_to_qobj(x)
+    elif isinstance(x, SparseDIAQArray):
+        return _sparsedia_to_qobj(x)
+    elif isinstance(x, Sequence) and all(isinstance(sub_x, QArray) for sub_x in x):
+        # TODO: generalize to any nested sequence with the appropriate shape
+        return [to_qutip(sub_x, dims=dims) for sub_x in x]
+
+    x = jnp.asarray(x)
+    check_shape(x, 'x', '(..., n, 1)', '(..., 1, n)', '(..., n, n)')
+    dims = _init_dims(x, dims)
+    dims = _dims_to_qutip(dims, x.shape)
+    return Qobj(x, dims=dims)
+
+
+def _warn_qarray_dims(x: QArrayLike, dims: tuple[int, ...] | None = None):
+    if dims is not None:
+        if isinstance(x, QArray) and x.dims != dims:
+            warnings.warn(
+                f'Argument `x` is already a QArray with dims={x.dims}, but dims '
+                f'were also provided as input with dims={dims}. Ignoring the '
+                'provided `dims` and proceeding with `x.dims`.',
+                stacklevel=2,
+            )
+        elif isinstance(x, Qobj) and _dims_from_qutip(x.dims) != dims:
+            warnings.warn(
+                f'Argument `x` is already a Qobj with dims={x.dims}, but dims '
+                f'were also provided as input with dims={dims}. Ignoring the '
+                'provided `dims` and proceeding with `x.dims`.',
+                stacklevel=2,
+            )
+
+
+def sparsedia_from_dict(
+    offsets_diags: dict[int, ArrayLike],
+    dims: tuple[int, ...] | None = None,
+    dtype: DTypeLike | None = None,
+) -> SparseDIAQArray:
+    """Initialize a `SparseDIAQArray` from a dictionary of offsets and diagonals.
+
+    Args:
+        offsets_diags: Dictionary where keys are offsets and values are diagonals of
+            shapes _(..., n-|offset|)_ with a common batch shape between all diagonals.
+        dims _(tuple of ints or None)_: Dimensions of each subsystem in the Hilbert
+            space tensor product. Defaults to `None` (a single system with the same
+            dimension as `x`).
+        dtype: Data type of the array. If `None`, the data type is inferred from the
+            diagonals.
+
+    Returns:
+        A `SparseDIAQArray` with non-zero diagonals at the specified offsets.
+
+    Examples:
+        >>> dq.sparsedia_from_dict({0: [1, 2, 3], 1: [4, 5], -1: [6, 7]})
+        QArray: shape=(3, 3), dims=(3,), dtype=int32, layout=dia, ndiags=3
+        [[1.+0.j 4.+0.j   ⋅   ]
+         [6.+0.j 2.+0.j 5.+0.j]
+         [  ⋅    7.+0.j 3.+0.j]]
+        >>> dq.sparsedia_from_dict({0: jnp.ones((3, 2))})
+        QArray: shape=(3, 2, 2), dims=(2,), dtype=float32, layout=dia, ndiags=1
+        [[[1.+0.j   ⋅   ]
+          [  ⋅    1.+0.j]]
+        <BLANKLINE>
+         [[1.+0.j   ⋅   ]
+          [  ⋅    1.+0.j]]
+        <BLANKLINE>
+         [[1.+0.j   ⋅   ]
+          [  ⋅    1.+0.j]]]
+    """
```

Comment:
There's something fishy with the indicated and actual types here.

Reply:
It's related to the question above, on whether we want to enforce the dtype to be complex or not. For now, `sparsedia_from_dict` is NOT enforcing the dtype, but `to_jax()` (called by `SparseDIAQArray.__repr__`) is, hence the mismatch.